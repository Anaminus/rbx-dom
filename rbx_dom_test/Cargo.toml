[package]
name = "rbx_dom_test"
version = "0.1.0"
authors = ["Lucien Greathouse <me@lpghatguy.com>"]
edition = "2018"
publish = false

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
rbx_dom_weak = { path = "../rbx_dom_weak" }
serde = "1.0"
serde_json = "1.0"

[dependencies.insta]
version = "0.12.0"
features = ["redactions"]
