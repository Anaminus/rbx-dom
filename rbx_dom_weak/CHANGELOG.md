# rbx\_dom\_weak Changelog

## Unreleased
* Renamed crate from `rbx_tree` to `rbx_dom_weak`
* Added support for `Ref` values ([#8](https://github.com/LPGhatguy/rbx-dom/pull/8))

## 0.2.0 (2019-01-25)
* Added new variants for `RbxValue`:
	* Int32
	* Float32
	* Enum
	* Vector2
	* Color3
	* Color3uint8
	* Vector3int16
	* Vector2int16
	* CFrame
	* PhysicalProperties (Stub)

## 0.1.0
* Initial release
* Supports `String`, `Bool`, and `Vector3` property values