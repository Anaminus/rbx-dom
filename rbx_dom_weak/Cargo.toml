[package]
name = "rbx_dom_weak"
description = "Weakly-typed Roblox DOM implementation for Rust"
version = "1.10.1"
license = "MIT"
documentation = "https://docs.rs/rbx_dom_weak"
homepage = "https://github.com/rojo-rbx/rbx-dom"
repository = "https://github.com/rojo-rbx/rbx-dom.git"
readme = "README.md"
authors = ["Lucien Greathouse <me@lpghatguy.com>"]
edition = "2018"

[dependencies]
rbx_types = { path = "../rbx_types", features = ["serde"] }

serde = "1.0"
serde_derive = "1.0"
serde_repr = "0.1"

[dev-dependencies]
serde_json = "1.0"
maplit = "1.0"
