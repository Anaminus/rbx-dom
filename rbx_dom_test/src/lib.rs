use std::collections::{BTreeMap, HashMap};

use rbx_dom_weak::{RbxId, RbxTree, RbxValue};
use serde::{Deserialize, Serialize};

/// Contains state for viewing/redacting RbxTree objects, making them suitable
/// for viewing in a snapshot test.
pub struct TreeViewer {
    id_map: HashMap<RbxId, String>,
    next_id: usize,
}

impl TreeViewer {
    pub fn new() -> Self {
        Self {
            id_map: HashMap::new(),
            next_id: 0,
        }
    }

    pub fn view(&mut self, tree: &RbxTree) -> ViewedInstance {
        let root_id = tree.get_root_id();
        self.populate_id_map(tree, root_id);
        self.view_instance(tree, root_id)
    }

    fn populate_id_map(&mut self, tree: &RbxTree, id: RbxId) {
        self.id_map.insert(id, format!("id-{}", self.next_id));
        self.next_id += 1;

        let instance = tree.get_instance(id).unwrap();
        for id in instance.get_children_ids() {
            self.populate_id_map(tree, *id);
        }
    }

    fn view_instance(&self, tree: &RbxTree, id: RbxId) -> ViewedInstance {
        let instance = tree.get_instance(id).unwrap();

        let children = instance
            .get_children_ids()
            .iter()
            .copied()
            .map(|id| self.view_instance(tree, id))
            .collect();

        let properties = instance
            .properties
            .iter()
            .map(|(key, value)| {
                let key = key.clone();
                let new_value = match value {
                    RbxValue::Ref {
                        value: Some(ref_id),
                    } => {
                        let id_str = self
                            .id_map
                            .get(ref_id)
                            .cloned()
                            .unwrap_or_else(|| "[unknown ID]".to_owned());
                        ViewedValue::Ref(id_str)
                    }
                    other => ViewedValue::Other(other.clone()),
                };

                (key, new_value)
            })
            .collect();

        ViewedInstance {
            id: self.id_map.get(&id).unwrap().clone(),
            name: instance.name.clone(),
            class_name: instance.class_name.clone(),
            properties,
            children,
        }
    }
}

/// A transformed view into an RbxTree or RbxInstance that has been redacted and
/// transformed to be more readable.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ViewedInstance {
    id: String,
    name: String,
    class_name: String,
    properties: BTreeMap<String, ViewedValue>,
    children: Vec<ViewedInstance>,
}

/// Wrapper around RbxValue with refs replaced to be redacted, stable versions
/// of their original IDs.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum ViewedValue {
    Ref(String),
    Other(RbxValue),
}

#[cfg(test)]
mod test {
    use super::*;

    use rbx_dom_weak::RbxInstanceProperties;

    #[test]
    fn redact_single() {
        let tree = RbxTree::new(RbxInstanceProperties {
            name: "Root".to_owned(),
            class_name: "Folder".to_owned(),
            properties: HashMap::new(),
        });

        insta::assert_yaml_snapshot!(TreeViewer::new().view(&tree));
    }

    #[test]
    fn redact_multi() {
        let mut tree = RbxTree::new(RbxInstanceProperties {
            name: "Root".to_owned(),
            class_name: "Folder".to_owned(),
            properties: HashMap::new(),
        });

        let root_id = tree.get_root_id();

        for i in 0..4 {
            let name = format!("Child {}", i);
            let properties = RbxInstanceProperties {
                name,
                class_name: "Folder".to_owned(),
                properties: HashMap::new(),
            };

            tree.insert_instance(properties, root_id);
        }

        insta::assert_yaml_snapshot!(TreeViewer::new().view(&tree));
    }

    #[test]
    fn redact_values() {
        let mut tree = RbxTree::new(RbxInstanceProperties {
            name: "Root".to_owned(),
            class_name: "ObjectValue".to_owned(),
            properties: HashMap::new(),
        });

        let root_id = tree.get_root_id();
        let root_instance = tree.get_instance_mut(root_id).unwrap();

        root_instance.properties.insert(
            "Value".to_owned(),
            RbxValue::Ref {
                value: Some(root_id),
            },
        );

        insta::assert_yaml_snapshot!(TreeViewer::new().view(&tree));
    }
}
