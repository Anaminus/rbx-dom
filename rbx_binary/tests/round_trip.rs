use std::{collections::HashMap, fs, io::Cursor, path::Path};

use log::trace;

use rbx_dom_test::TreeViewer;
use rbx_dom_weak::{RbxInstanceProperties, RbxTree};

#[test]
fn just_folder() {
    test_round_trip("just-folder.rbxm");
}

#[test]
fn folder_with_script() {
    test_round_trip("folder-with-script.rbxm");
}

#[test]
fn hierarchy_with_refs() {
    test_round_trip("hierarchy-with-refs.rbxm");
}

/// Helper method to test both reading and writing using only rbx_binary.
fn test_round_trip(test_name: &str) {
    let _ = env_logger::try_init();

    let input_path = Path::new("test-files").join(test_name);
    let initial_contents = fs::read(&input_path).unwrap();

    let mut initial_tree = new_test_tree();

    let root_id = initial_tree.get_root_id();
    rbx_binary::decode(&mut initial_tree, root_id, initial_contents.as_slice()).unwrap();

    let initial_viewed = TreeViewer::new().view_children(&initial_tree);
    insta::assert_yaml_snapshot!(test_name, initial_viewed);

    let root_instance = initial_tree.get_instance(root_id).unwrap();
    let root_children = root_instance.get_children_ids();
    let mut roundtrip_contents = Vec::new();
    rbx_binary::encode(&initial_tree, root_children, &mut roundtrip_contents).unwrap();

    let mut roundtrip_tree = new_test_tree();

    let root_id = roundtrip_tree.get_root_id();
    rbx_binary::decode(&mut roundtrip_tree, root_id, roundtrip_contents.as_slice()).unwrap();

    let roundtrip_viewed = TreeViewer::new().view_children(&roundtrip_tree);
    insta::assert_yaml_snapshot!(test_name, roundtrip_viewed);
}

fn new_test_tree() -> RbxTree {
    let root = RbxInstanceProperties {
        name: "Folder".to_string(),
        class_name: "Folder".to_string(),
        properties: HashMap::new(),
    };

    RbxTree::new(root)
}
