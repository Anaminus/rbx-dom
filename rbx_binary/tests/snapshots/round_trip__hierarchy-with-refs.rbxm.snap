---
source: rbx_binary/tests/round_trip.rs
expression: initial_viewed
---
- id: id-0
  name: A
  class_name: Folder
  properties:
    AttributesSerialize:
      Type: BinaryString
      Value: ""
    Tags:
      Type: BinaryString
      Value: ""
  children:
    - id: id-1
      name: B
      class_name: Folder
      properties:
        AttributesSerialize:
          Type: BinaryString
          Value: ""
        Tags:
          Type: BinaryString
          Value: ""
      children:
        - id: id-2
          name: C
          class_name: Part
          properties:
            Anchored:
              Type: Bool
              Value: false
            AttributesSerialize:
              Type: BinaryString
              Value: ""
            CanCollide:
              Type: Bool
              Value: true
            CastShadow:
              Type: Bool
              Value: true
            Locked:
              Type: Bool
              Value: false
            Massless:
              Type: Bool
              Value: false
            Tags:
              Type: BinaryString
              Value: ""
          children:
            - id: id-3
              name: Weld
              class_name: Weld
              properties:
                AttributesSerialize:
                  Type: BinaryString
                  Value: ""
                Enabled:
                  Type: Bool
                  Value: true
                IsAutoJoint:
                  Type: Bool
                  Value: false
                Tags:
                  Type: BinaryString
                  Value: ""
              children: []
